### PR TITLE
WIP: go2idl examples

### PR DESCRIPTION
**Pull request by lavalamp**
The not-fully-baked examples from #14665, in separate commits. Some commits from #15714.

### CODE REVIEW

**Comment by wojtek-t** [comment]
@lavalamp - since I hope that the core PR will be merge very soon - I would like to start reviewing this one.
However, I have some requests about it:
- can you please divide this PR into multiple so that there is one example per PR?
- into those PRs can you have one commit with the code written by you and one with the auto-generated one (I know it's already here)
Thanks!

**Comment by wojtek-t** [comment]
@lavalamp - I'm happy to help with making this PR happen after #15714 is merged - I can take either "client" or "deep-copy" from it - feel free to choose which one do you want me to work on.

**Comment by wojtek-t** [comment]
@lavalamp - I'm taking your "copy-gen" part of this PR and try to make it work. Will send a separate PR for it.

**Comment by lavalamp** [comment]
Thanks @wojtek-t! You may want to take a look here:
https://github.com/kubernetes/kubernetes/pull/16062/files#diff-b9c92f79197383b1ee5ac7c3261b579aR338
and leave a comment if what you want to do seems incompatible with what I'm
thinking over there.

File: cmd/libs/go2idl/client-gen/main.go [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/kubernetes/cmd/libs/go2idl/generator"
```
Comment (thockin) [inline cmd/libs/go2idl/client-gen/main.go]: Are all of these pkgs used together all the time?  I get the desire to split them out, but as I read over this they all "feel" like they are too inter-related to be distinct pkgs.

File: cmd/libs/go2idl/set-gen/main.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+	"k8s.io/kubernetes/cmd/libs/go2idl/parser"
+	"k8s.io/kubernetes/cmd/libs/go2idl/types"
+
+	"github.com/golang/glog"
+	flag "github.com/spf13/pflag"
+)
+
+var (
+	inputDirs         = flag.StringP("input-dirs", "i", "./types", "comma-separated list of directories to get input types from.")
+	outputBase        = flag.StringP("output-base", "o", filepath.Join(os.Getenv("GOPATH"), "src"), "Output base; defaults to $GOPATH/src/")
```
Comment (thockin) [inline cmd/libs/go2idl/set-gen/main.go]: not cool if GOPATH has multiple entries

Comment (thockin) [inline cmd/libs/go2idl/set-gen/main.go]: I'd argue for "dir" in the flagname somewhere

Comment (lavalamp) [inline cmd/libs/go2idl/set-gen/main.go]: ARGH I did that right in one place and forgot about it here

**Comment by wojtek-t** [comment]
I'm closing this PR, because:
- set-gen and copy-gen were already merged in different PRs
- client-gen is actively being developed by @caesarxuchao

**Pull request closed without merging**